# Hypothesis testing

### Definition: 
#### Hypothesis testing is a method of inferential statistic used to determine the possible conclusion which has to be  statistically proven or not.
  


In [1]:
!pip install scipy==1.6.0

In [2]:
from scipy import stats
import pandas as pd
import numpy as np
import math as mt
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

# Question 1
 A F&B manager wants to determine whether there is any significant difference in the diameter of the cutlet between two units. A randomly selected sample of cutlets was collected from both units and measured? Analyze the data and draw inferences at 5% significance level. Please state the assumptions and tests that you carried out to check validity of the assumptions.




##### Define the hypothesis :

##### H0 hypothesis => There is no difference in the diameter of the cutlets of Unit A and B,

##### Ha hypothesis => There is a significant difference in the diameter of the cutlets of Unit A and B



In [3]:
df = pd.read_csv('Cutlets.csv')

In [4]:
df.head(10)

Unit A  Unit B
0  6.8090  6.7703
1  6.4376  7.5093
2  6.9157  6.7300
3  7.3012  6.7878
4  7.4488  7.1522
5  7.3871  6.8110
6  6.8755  7.2212
7  7.0621  6.6606
8  6.6840  7.2402
9  6.8236  7.0503

Since we have sample data we will go for t-test to compare the means of two samples.

In [5]:
stats.ttest_ind(df['Unit A'],df['Unit B'])

Ttest_indResult(statistic=0.7228688704678063, pvalue=0.47223947245995)


##### pvalue > alpha (0.4722 > 0.05)

##### we will fail to reject the Null hypothesis here

##### Accept H0 => There is no difference in the diameter of the cutlets of Unit A and B

# Question 2

 A hospital wants to determine whether there is any difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list. They collected a random sample and recorded TAT for reports of 4 laboratories. TAT is defined as sample collected to report dispatch. Analyze the data and determine whether there is any difference in average TAT among the different laboratories at 5% significance level.


##### H0 hypothesis: The male-female buyer ratios are similar across the regions / All proportions are equal

##### Ha hypothesis:The male-female buyer ratios are not similar across the regions / Not all proportions are equal

In [6]:
import scipy.stats as stats

In [7]:
labtat = pd.read_csv('LabTaT.csv')

In [8]:
labtat.head(10)

Laboratory 1  Laboratory 2  Laboratory 3  Laboratory 4
0        185.35        165.53        176.70        166.13
1        170.49        185.91        198.45        160.79
2        192.77        194.92        201.23        185.18
3        177.33        183.00        199.61        176.42
4        193.41        169.57        204.63        152.60
5        179.45        197.00        181.51        161.12
6        191.37        166.36        214.21        154.02
7        166.81        169.60        183.43        163.25
8        158.81        175.36        191.60        152.79
9        165.88        198.68        208.43        161.98

In [9]:
labtat.describe()

Laboratory 1  Laboratory 2  Laboratory 3  Laboratory 4
count    120.000000    120.000000    120.000000     120.00000
mean     178.361583    178.902917    199.913250     163.68275
std       13.173594     14.957114     16.539033      15.08508
min      138.300000    140.550000    159.690000     124.06000
25%      170.335000    168.025000    188.232500     154.05000
50%      178.530000    178.870000    199.805000     164.42500
75%      186.535000    189.112500    211.332500     172.88250
max      216.390000    217.860000    238.700000     205.18000

The F-Test is utilized to determine whether two populations' variances are equal.

In [10]:
stats.f_oneway(labtat.iloc[:,0],labtat.iloc[:,1],labtat.iloc[:,2],labtat.iloc[:,3])


F_onewayResult(statistic=118.70421654401437, pvalue=2.1156708949992414e-57)

##### pvalue > alpha (2.115>0.05)

##### so choose null hpothesis and reject alternate hypothesis.

##### Therefore, the male-female buyer ratios are similar across the regions / All proportions are equal.

# Question 3
Sales of products in four different regions is tabulated for males and females. Find if male-female buyer ratios are similar across regions.


##### H0: The male/female buyer ratio are similar 

##### H1: The male/female buyer ratio are not similar 

In [11]:
buyer = pd.read_csv('BuyerRatio.csv')
buyer

Observed Values  East  West  North  South
0           Males    50   142    131     70
1         Females   435  1523   1356    750

In [12]:
import warnings
warnings.filterwarnings("ignore")


In [13]:
buyer= pd.DataFrame([
    [50, 142, 131, 70],
    [435, 1523, 1356, 750]], 
    index = ['Male', 'Female'], columns = ['East', 'West', 'North', 'South'])
print(buyer)

        East  West  North  South
Male      50   142    131     70
Female   435  1523   1356    750


we choose chi square test to examine the differences between categorical variables from a random sample in order to judge goodness of fit between expected and observed results.

In [14]:
from scipy.stats import chi2_contingency
  
stat, p, dof, expected = chi2_contingency(buyer)
print(stat,"\n", p,"\n", dof,"\n", expected)


1.595945538661058 
 0.6603094907091882 
 3 
 [[  42.76531299  146.81287862  131.11756787   72.30424052]
 [ 442.23468701 1518.18712138 1355.88243213  747.69575948]]


In [15]:
observed = np.array([50, 142, 131, 70, 435, 1523, 1356, 750])
expected = np.array([42.76531299,  146.81287862,  131.11756787, 72.30424052, 442.23468,1518.18712138, 1355.88243213, 747.69575948])

In [16]:
statistics, p_value = stats.chisquare(observed, expected, ddof = 3)
p_value

0.8095207055554905

#####  pvalue > alpha (0.809 > 0.05)

##### Hence we will fail to reject H0 (Null hypothesis) here.


# Question 4
TeleCall uses 4 centers around the globe to process customer order forms. They audit a certain % of the customer order forms. Any error in order form renders it defective and has to be reworked before processing. The manager wants to check whether the defective % varies by centre. Please analyze the data at 5% significance level and help the manager draw appropriate inferences

##### H0 hypothesis: the defective % does not vary by centre / the variables are independent 
 
##### Ha hypothesis: the defective % varies by centre / the variables are dependent


In [17]:
centers = pd.read_csv('Costomer+OrderForm.csv')
centers.head(5)

Phillippines   Indonesia       Malta       India
0   Error Free  Error Free   Defective  Error Free
1   Error Free  Error Free  Error Free   Defective
2   Error Free   Defective   Defective  Error Free
3   Error Free  Error Free  Error Free  Error Free
4   Error Free  Error Free   Defective  Error Free

In [18]:
centers.describe()

Phillippines   Indonesia       Malta       India
count           300         300         300         300
unique            2           2           2           2
top      Error Free  Error Free  Error Free  Error Free
freq            271         267         269         280

In [19]:
centers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Phillippines  300 non-null    object
 1   Indonesia     300 non-null    object
 2   Malta         300 non-null    object
 3   India         300 non-null    object
dtypes: object(4)
memory usage: 9.5+ KB


In [20]:
print(centers['Phillippines'].value_counts(),'\n',centers['Indonesia'].value_counts(),'\n',centers['Malta'].value_counts(),centers['India'].value_counts())

Error Free    271
Defective      29
Name: Phillippines, dtype: int64 
 Error Free    267
Defective      33
Name: Indonesia, dtype: int64 
 Error Free    269
Defective      31
Name: Malta, dtype: int64 Error Free    280
Defective      20
Name: India, dtype: int64


In [21]:
ctable = [[271,267,269,280],
                    [29,33,31,20]]
print(ctable)

[[271, 267, 269, 280], [29, 33, 31, 20]]


In [22]:
stat, p, df, exp = stats.chi2_contingency(ctable)
print("Statistics = ",stat,"\n",'p_value= ',p,'\n', 'degree of freedom =', df,'\n','Expected Values=',exp)

Statistics =  3.858960685820355 
 p_value=  0.2771020991233135 
 degree of freedom = 3 
 Expected Values= [[271.75 271.75 271.75 271.75]
 [ 28.25  28.25  28.25  28.25]]


In [23]:
observed = np.array([271, 267, 269, 280, 29, 33, 31, 20])
expected = np.array([271.75, 271.75, 271.75, 271.75, 28.25, 28.25, 28.25, 28.25])

In [24]:
test_statistics, p_value = stats.chisquare(observed, expected, ddof = 3)
p_value

0.4254298144535761

#####  Final Statement : Now if we analyse pvalue = 0.277 at 5% significance level (alpha = 0.05) we get pvlaue greater than alpha. Hence we will fail to reject H0 (Null Hypothesis)

#####  pvalue > alpha (0.425 > 0.05)